pkg/orchestrator/worker.go

```diff
@@ -0,0 +1,197 @@
+	"github.com/rs/zerolog/log"
+	"go.uber.org/atomic"
+)
+
+const (
+	WorkerStatusInit     = "Initialized"
+	WorkerStatusStarting = "Starting"
+	WorkerStatusRunning  = "Running"
+	WorkerStatusStopping = "Stopping"
+	WorkerStatusStopped  = "Stopped"
+)
+
+type WorkerParams struct {
+	// SchedulerProvider is responsible for providing the scheduler instance
+	// based on the evaluation type.
+	SchedulerProvider SchedulerProvider
+	// EvaluationBroker is the broker used for handling evaluations.
+	EvaluationBroker EvaluationBroker
+	// DequeueTimeout is the maximum duration for dequeueing an evaluation.
+	DequeueTimeout time.Duration
+	// DequeueFailureBackoff defines the backoff strategy when dequeueing an evaluation fails.
+	DequeueFailureBackoff backoff.Backoff
+}
+
+// Worker is a long-running process that dequeues evaluations, invokes the scheduler
+// to process the evaluation, and acknowledges or rejects the evaluation back to the broker.
+// The worker is single-threaded and processes one evaluation at a time. An orchestrator
+// can have multiple workers working in parallel.
+type Worker struct {
+	schedulerProvider     SchedulerProvider
+	evaluationBroker      EvaluationBroker
+	dequeueTimeout        time.Duration
+	dequeueFailureBackoff backoff.Backoff
+
+	status          atomic.String
+	startOnce       sync.Once
+	shutdownOnce    sync.Once
+	shuttingDownCtx context.Context
+	shuttingDownFn  context.CancelFunc
+}
+
+// NewWorker returns a new Worker instance.
+func NewWorker(ctx context.Context, params *WorkerParams) *Worker {
+	w := &Worker{
+		schedulerProvider:     params.SchedulerProvider,
+		evaluationBroker:      params.EvaluationBroker,
+		dequeueTimeout:        params.DequeueTimeout,
+		dequeueFailureBackoff: params.DequeueFailureBackoff,
+		status:                *atomic.NewString(WorkerStatusInit),
+	}
+	w.shuttingDownCtx, w.shuttingDownFn = context.WithCancel(ctx)
+	return w
+}
+
+// Start triggers the worker to start processing evaluations.
+// The worker can only start once, and subsequent calls to Start will be ignored.
+func (w *Worker) Start() {
+	w.startOnce.Do(func() {
+		w.setStatus(WorkerStatusStarting)
+		go w.run()
+	})
+}
+
+// Stop triggers the worker to stop processing evaluations.
+// The worker will stop after the in-flight evaluation is processed.
+func (w *Worker) Stop() {
+	w.shutdownOnce.Do(func() {
+		w.setStatus(WorkerStatusStopping)
+		w.shuttingDownFn()
+	})
+}
+
+// Status returns the current status of the worker.
+func (w *Worker) Status() string {
+	return w.status.Load()
+}
+
+func (w *Worker) run() {
+	defer w.setStatus(WorkerStatusStopped)
+	w.setStatus(WorkerStatusRunning)
+
+	var dequeueFailures int
+	for !w.isShuttingDown() {
+		// Dequeue an evaluation and apply backoff if dequeueing fails
+		evaluationReceipt, err := w.dequeueEvaluation()
+		if err != nil {
+			dequeueFailures++
+			w.dequeueFailureBackoff.Backoff(w.shuttingDownCtx, dequeueFailures)
+			continue
+		}
```

Comment:
It might be worth exploring [sync.Cond](https://pkg.go.dev/sync#NewCond) in a follow for synchronizing dequeue events from the evaluation broker.

Reply:
Looks useful. Thanks for sharing